[Mobile Payments] Redirect Simple Payments from Order navigation to HubMenu

iamgabrielma opened the pull request:
Closes: #7440 

### Description

This PR changes the navigation flow when a merchant attempts to create a `Simple Payment` within the Orders screen. Rather than being redirected to the Simple Payments flow opener, now they're shown a notice that informs them that the feature has been moved to the app's Menu tab (HubMenu). Upon tapping on the `Got it!`button, they're redirected to the app's Menu tab.

### Changes

If the feature flag is active, we intercept the current `presentOrderCreationFlow` and show a `BottomSheetViewController` in our `AddOrderCoordinator`. This bottom sheet is an existing WordPressUI component that already implements the base of what we need for displaying a modal view on-screen.

Through this modal, as the Order screen is done in UIKit, we use a `UIHostingViewController` which wraps a new reusable SwiftUI `BottomSheetNoticeView` View. Behaviour and configuration for this is done via the model `NewSimplePaymentsNoticeViewModel`.


### Testing instructions
1. Go to Orders > Tap "+" > Tap "Simple Payments" > See the new bottom sheet appear.
2. Tap in "Got it!" > Confirm you're redirected to the HubMenu

### Screenshots

| Light | Dark |
|--|--|
| <img width=400 src="https://user-images.githubusercontent.com/3812076/184135355-a3b43be3-d9f9-4adc-93fb-698cce5bb62c.png" > | <img width=400 src="https://user-images.githubusercontent.com/3812076/184135376-5348c5a1-259a-4a4c-a48e-691edf251399.png"> |

https://user-images.githubusercontent.com/3812076/184136336-f024b89c-1ade-4ee6-9b99-dce3d02d18a9.mp4


---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

iamgabrielma committed:
Add base UIHostingController to coordinator

iamgabrielma committed:
Create BottomAnnouncementView View

iamgabrielma committed:
Add BottomAnnouncementView dismiss

iamgabrielma committed:
Add mainTabBar navigation switch

iamgabrielma committed:
Wrapped View into hostingController

iamgabrielma committed:
Add callback completion handler

iamgabrielma committed:
Add featureflag filter

iamgabrielma committed:
Style BottomAnnouncement View

iamgabrielma committed:
Remove unwanted file

iamgabrielma committed:
Add guard. Move setupConstraints to its own method

iamgabrielma committed:
Improve completionHandler naming

iamgabrielma committed:
Add bottom constraints to controller

iamgabrielma committed:
Add heigh constraints to controller

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/FlowCoordinator/AddOrderCoordinator.swift:
```diff
@@ -68,6 +73,23 @@ private extension AddOrderCoordinator {
+    ///
+    func presentBottomAnnouncement() {
+        let baseViewController = BaseViewController()
+        baseViewController.completionHandler = redirectToHubMenu
```
As talked this can be added into a View Model

iamgabrielma replied:
Done at 94997e002f6e4f4c90a6260b45fa98f8f4d30157 👍

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/FlowCoordinator/AddOrderCoordinator.swift:
```diff
@@ -86,3 +108,48 @@ private extension AddOrderCoordinator {
     }
 }
+
+class BaseViewController: UIViewController {
```
We can find a more meaningful name for this class

iamgabrielma replied:
Definitely! Renamed at aee2e57f3b3ae77dc7c6a4acffda42997787d7d0

toupper commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/BottomAnnouncementView.swift:
```diff
@@ -0,0 +1,57 @@
+
+extension BottomAnnouncementView {
+    enum Localization {
+        static let title = NSLocalizedString("Payments from the Menu tab",
```
If we create a view model we could pass these strings through the initializer, so this view can remain totally generic.

iamgabrielma replied:
Great idea, done here: 5a8abf6190078608d804c1c88dd9d4a50067ea18

toupper committed:
Pairing session tweaks

iamgabrielma committed:
Renamed View and ViewController for specificity

iamgabrielma committed:
Add inline comments

iamgabrielma committed:
Design changes to Announcement sheet

iamgabrielma committed:
Create view model

iamgabrielma committed:
Extract viewmodel to its own file

iamgabrielma committed:
More meaningful variable naming

iamgabrielma committed:
Make classes final, and methods private

iamgabrielma committed:
Made View generic. Move strings to viewModel

iamgabrielma committed:
Merge branch 'trunk' into issue/7440-change-navigation-ipp-orders

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/FlowCoordinator/AddOrderCoordinator.swift:
```diff
@@ -86,3 +107,34 @@ private extension AddOrderCoordinator {
 }
+
+final class NewSimplePaymentsNoticeViewController: UIViewController {
+
```
We don't need this extra empty line here, same below

iamgabrielma replied:
Cleaned up here 5ab165eb82d811157aaf9f236ed898662eb82fd1

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/FlowCoordinator/NewSimplePaymentsNoticeViewModel.swift:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    func setupNewSimplePaymentsNoticeView(for viewController: UIViewController) {
+        let hostingController = UIHostingController(rootView: simplePaymentsNoticeView)
```
This is view related logic, that should not be in the view model actually. We can move this into the `UIViewController` :)

iamgabrielma replied:
Gotcha! Also thanks for the further explanation during our session, I moved the view-related logic to the ViewController and extracted it to its own file: e128e1e31d27efa2dee2138d6a1c4d584d245ee8

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/FlowCoordinator/NewSimplePaymentsNoticeViewModel.swift:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    private func setupConstraints(for hostingController: UIHostingController<BottomSheetNoticeView>, with viewController: UIViewController) {
+        hostingController.view.translatesAutoresizingMaskIntoConstraints = false
```
Same for this, this view logic should be in the view controller.

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/FlowCoordinator/NewSimplePaymentsNoticeViewModel.swift:
```diff
@@ -0,0 +1,54 @@
+
+    /// Redirects to `HubMenu`tabBar
+    ///
+    private var completionHandler: (() -> Void)? = {
```
I find the `completionHandler` naming a bit confusing here, there are two ways of completion here, dismissing by canceling and navigating to menu. This closure only takes the latter into account, so I would rename it to something more specific, like `navigateToMenuButtonWasTapped`

iamgabrielma replied:
Makes sense! Changed here: fdf13a636fd3bd800ba71ad302d42e0984d43981

toupper commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/BottomSheetNoticeView.swift:
```diff
@@ -0,0 +1,55 @@
+import SwiftUI
+
+struct BottomSheetNoticeView: View {
```
Nice generic view! Perhaps we can make the name more generic, after all we can use it in scenarios other than bottom and sheet. Something like `DismissableNoticeView` ?

iamgabrielma replied:
Definitely, done: ad4287c2a805da29e0e1170b277214407e1ee09d

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/FlowCoordinator/AddOrderCoordinator.swift:
```diff
@@ -86,3 +107,34 @@ private extension AddOrderCoordinator {
     }
 }
+
+final class NewSimplePaymentsNoticeViewController: UIViewController {
```
Nice naming! I would add something to let it know that Simple Payments are now in a different app place, otherwise, it seems like we are using a new simple payments feature. Something like `NewSimplePaymentsLocationNoticeViewController` ?

iamgabrielma replied:
Done! 22da3170258ac0e5eebdcbeff75e1c98a5098fc5

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/FlowCoordinator/NewSimplePaymentsNoticeViewModel.swift:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    init() {
+        simplePaymentsNoticeView = BottomSheetNoticeView(
```
This view logic should also go to the view controller in my opinion. From there it can access the localization strings, that can stay here in the view model :)

iamgabrielma replied:
Gotcha! I moved the init into the ViewController here: 9c9b1222af97cb6bb0b278996c1396e0a17b3953 , so now the ViewModel only has the button tapped logic, the icon data, and the localisation strings 👍

toupper commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/BottomSheetNoticeView.swift:
```diff
@@ -0,0 +1,55 @@
+import SwiftUI
+
```
A comment about what this view is doing would be nice :)

iamgabrielma replied:
Fixed! acd012e9d40236ed00c1738c0aec89ecc21ad49e

toupper commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/BottomSheetNoticeView.swift:
```diff
@@ -0,0 +1,55 @@
+                    .multilineTextAlignment(.leading)
+                    .fixedSize(horizontal: false, vertical: true)
+                Spacer()
+                Image(uiImage: .walletImage)
```
This image is hardcoded, we can make it injectable from the outside to keep it generic :)

iamgabrielma committed:
removed empty lines

iamgabrielma committed:
renamed DismissableNoticeView for clarity

iamgabrielma committed:
renamed completion handler for clarity

iamgabrielma committed:
make UIImage injectable

iamgabrielma committed:
Add comment to DismissableNoticeView

iamgabrielma committed:
Rename NewSimplePaymentsLocationNoticeViewController

iamgabrielma committed:
Extracted view logic to view controller

iamgabrielma committed:
Moved init View logic to view controller

iamgabrielma committed:
Updated comment after renaming class

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/FlowCoordinator/NewSimplePaymentsLocationNoticeViewModel.swift:
```diff
@@ -0,0 +1,40 @@
+    /// Uses the "woo-payments-wallet" UIImage+Woo as icon
+    ///
+    var icon: UIImage {
+        guard let image = UIImage(named: "woo-payments-wallet") else { return UIImage() }
```
You can use here the `static var walletImage: UIImage ` in UIImage+Woo, and avoid the guard

iamgabrielma replied:
That makes sense, changed at 40d089712c45a0bfb74ad2298cddc5f5403212c0

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/FlowCoordinator/NewSimplePaymentsLocationNoticeViewModel.swift:
```diff
@@ -0,0 +1,40 @@
+    let message: String
+    let confirmationButtonMessage: String
+
+    init() {
```
It would be nice if we follow here the usual order for a class elements:

- stored properties
- computed properties
- init
- public functions
- private functions (in a private extension)

toupper committed:
Merge branch 'trunk' into issue/7440-change-navigation-ipp-orders

toupper commented:
The action looks good to me! I just added a couple of small code remarks. :) 
Besides that, it seems to me that the implemented layout does not fully match the design attached to the issue. Please take a look at this annotated screenshot, you can ask designers for full feedback :) 

![Annotation on 2022-08-12 at 16-43-56](https://user-images.githubusercontent.com/1864060/184379930-23093be8-b6e2-4dd9-87a0-02ed923f66dc.png)

iamgabrielma committed:
Adapt view to designs. Add View PreviewProvider

iamgabrielma committed:
Use static UIImage+Woo. Reorganize var declaration

iamgabrielma committed:
Merge branch 'issue/7440-change-navigation-ipp-orders' of https://github.com/woocommerce/woocommerce-ios into issue/7440-change-navigation-ipp-orders

iamgabrielma committed:
Removed DissmissableNoticeView extra padding

iamgabrielma commented:
> Besides that, it seems to me that the implemented layout does not fully match the design attached to the issue. 

Oh, you're right! I didn't see we have a Figma design for the i1 version so I implemented the layout by eye following the gif 😅 I've fixed the view following the design, changes at c0cd8027bfa5e2298ced6f8f1ca98ffe619fd3e2 . 

Here are the new screenshots:

| Light Mode | Dark Mode |
|--|--|
|<img width=350 src="https://user-images.githubusercontent.com/3812076/184532745-6abfff68-080a-49fc-b2f9-7ffd4c4d41e7.png">|<img width=350 src="https://user-images.githubusercontent.com/3812076/184532746-63e4ecc9-14b4-4dc1-942c-3f807a4f632f.png">|

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/FlowCoordinator/NewSimplePaymentsLocationNoticeViewController.swift:
```diff
@@ -0,0 +1,62 @@
+        super.viewWillLayoutSubviews()
+    }
+
+    func setupNewSimplePaymentsNoticeView() {
```
Not a blocker, but this could be `private` if it is not called externally

iamgabrielma committed:
Make method private. Change var for let
